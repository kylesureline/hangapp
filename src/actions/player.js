import database from '../firebase/firebase';

export const initializeGame = (word, guessesRemaining) => ({
  type: 'INIT_GAME',
  answer: word,
  guessesRemaining
});

export const guessLetter = (letter) => ({
  type: 'GUESS_LETTER',
  letter
});

export const updateGuessedWord = (wordArray) => ({
  type: 'UPDATE_GUESSED_WORD',
  wordArray
});

export const wrongGuess = () => ({
  type: 'WRONG_GUESS'
});

export const setPlayer = (player) => ({
  type: 'SET_PLAYER',
  player
});

export const setPlayerStats = (stats) => ({
  type: 'SET_PLAYER_STATS',
  stats
});

export const setPastGames = (pastGames) => ({
  type: 'SET_PAST_GAMES',
  pastGames
});

export const startSetPlayer = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`players/${uid}`)
      .once('value')
      .then((snapshot) => {
        const pastGames = [];

        for(let game in snapshot.val().pastGames) {
          pastGames.push(snapshot.val().pastGames[game]);
        }

        const stats = snapshot.val().stats;

        dispatch(setPlayerStats(stats));
        dispatch(setPastGames(pastGames));
      });

  };
};
