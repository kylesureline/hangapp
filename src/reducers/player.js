const defaultGameState = {
  answer: {
    word: '',
    type: '',
    def: ''
  },
  guessedLetters: [],
  guessedWord: [],
  guessesRemaining: 0,
  stats: {
    wins: 0,
    losses: 0
  }
};

export default (state = defaultGameState, action) => {
  switch(action.type) {
    case 'GUESS_LETTER':
      return {
        ...state,
        guessedLetters: [...state.guessedLetters, action.letter]
      };
    case 'UPDATE_GUESSED_WORD':
      return {
        ...state
      };
    case 'WRONG_GUESS':
      return {
        ...state,
        guessesRemaining: state.guessesRemaining -= 1
      };
    case 'SET_PLAYER_STATS':
      return {
        ...state,
        stats: action.stats
      };
    case 'SET_PAST_GAMES':
      return {
        ...state,
        pastGames: action.pastGames
      };
    case 'SET_GUESSES_REMAINING':
      return {
        ...state,
        guessesRemaining: action.guessesRemaining
      }
    case 'CHOOSE_DEFAULT_WORD':
      return {
        ...state,
        answer: {
          word: 'hangapp',
          type: 'noun',
          def: 'A hangman game built as a React PWA by Kyle Scheuerlein'
        }
      };
    case 'CHOOSE_RANDOM_WORD':
      return {
        ...state,
        answer: action.answer
      }
    case 'ADD_WORD':
      return {
        ...state,
        pastGames: [
          ...state.pastGames,
          action.answer
        ]
      }
    case 'ADD_WIN':
      return {
        ...state,
        stats: {
          ...state.stats,
          wins: state.stats.wins + 1
        }
      }
    case 'ADD_LOSS':
      return {
        ...state,
        stats: {
          ...state.stats,
          losses: state.stats.losses + 1
        }
      }
    default:
      return state;
  }
};
