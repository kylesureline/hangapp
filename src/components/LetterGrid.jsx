import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import KeyboardEventHandler from 'react-keyboard-event-handler';
import { guessLetter, updateGuessedWord, wrongGuess } from '../actions/player';
import LetterGridItem from './LetterGridItem.jsx';
import { startAddWord, startAddWin, startAddLoss } from '../actions/player';

export class LetterGrid extends React.Component {
  handleGuess = (e) => {
    const letter = e.target.textContent;
    if(!this.isGuessedLetter(letter)) {
      this.props.guessLetter(letter);
      this.checkWord(letter);
    }
  };
  handleKeyPress = (letter) => {
    if(!this.isGuessedLetter(letter)) {
      this.props.guessLetter(letter);
      this.checkWord(letter);
    }
  };
  isGuessedLetter = (letter) => {
    const guessedLetters = this.props.guessedLetters;
    if(guessedLetters.indexOf(letter) === -1) {
      return false;
    }
    return true;
  };
  checkWord = (letter) => {
    const answer = this.props.answer.word;
    const guessedWord = this.props.guessedWord;
    let wrong = 0;
    for(let i = 0; i < answer.length; i++) {
      if(letter === answer[i]) {
        guessedWord[i] = letter;
        this.props.updateGuessedWord(guessedWord);
      } else {
        wrong++;
      }
    }
    if(wrong === answer.length) {
      this.props.wrongGuess();
    }
    if(answer === guessedWord.join('') && this.props.guessesRemaining > 0) {
      this.props.startAddWin();
      this.props.startAddWord({...this.props.answer, won: true});
    }
    if((this.props.guessesRemaining - 1) <= 0) {
      this.props.startAddLoss();
      this.props.startAddWord({...this.props.answer, won: false});
    }
  };
  render() {
    return (
      <div>
        {
          'abcdefghijklmnopqrstuvwxyz'.split('').map((letter) => {
            return <LetterGridItem
                      key={letter}
                      letter={letter}
                      handleGuess={this.handleGuess}
                      handleKeyPress={this.handleKeyPress}
                    />;
          })
        }
        <KeyboardEventHandler
          handleKeys={'abcdefghijklmnopqrstuvwxyz'.split('')}
          onKeyEvent={this.handleKeyPress}
        />
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  answer: state.player.answer,
  guessedLetters: state.player.guessedLetters,
  guessesRemaining: state.player.guessesRemaining,
  guessedWord: state.player.guessedWord
});

const mapDispatchToProps = (dispatch) => ({
  guessLetter: (letter) => dispatch(guessLetter(letter)),
  updateGuessedWord: (guessedWord) => dispatch(updateGuessedWord(guessedWord)),
  wrongGuess: () => dispatch(wrongGuess()),
  startAddWord: (answer) => dispatch(startAddWord(answer)),
  startAddWin: () => dispatch(startAddWin()),
  startAddLoss: () => dispatch(startAddLoss())
});

export default connect(mapStateToProps, mapDispatchToProps)(LetterGrid);
