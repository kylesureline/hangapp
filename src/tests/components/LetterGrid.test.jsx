import React from 'react';
import { shallow } from 'enzyme';
import { LetterGrid } from '../../components/LetterGrid.jsx';
import { player } from '../fixtures/player';
let guessLetter,
    updateGuessedWord,
    wrongGuess,
    startAddWord,
    startAddWin,
    startAddLoss,
    answer,
    guessedLetters,
    guessesRemaining,
    guessedWord;

const playerWithNoGuesses = {
  ...player,
  guessedLetters: [],
  guessesRemaining: 10,
  guessedWord: []
};

const playerWithOneWrongGuess = {
  ...player,
  guessedLetters: ['z'],
  guessesRemaining: 9,
  guessedWord: []
}

beforeEach(() => {
  guessLetter = jest.fn();
  updateGuessedWord = jest.fn();
  wrongGuess = jest.fn();
  startAddWord = jest.fn();
  startAddWin = jest.fn();
  startAddLoss = jest.fn();
});

test('should render LetterGrid correctly', () => {
  const wrapper = shallow(
    <LetterGrid
      guessLetter={guessLetter}
      updateGuessedWord={updateGuessedWord}
      wrongGuess={wrongGuess}
      startAddWord={startAddWord}
      startAddWin={startAddWin}
      startAddLoss={startAddLoss}
      answer={playerWithNoGuesses.answer}
      guessedLetters={playerWithNoGuesses.guessedLetters}
      guessesRemaining={playerWithNoGuesses.guessesRemaining}
      guessedWord={playerWithNoGuesses.guessedWord}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
