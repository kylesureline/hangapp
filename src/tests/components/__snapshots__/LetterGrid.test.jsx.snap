// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LetterGrid correctly 1`] = `
<div
  className="letter-grid"
>
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="a"
    letter="a"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="b"
    letter="b"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="c"
    letter="c"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="d"
    letter="d"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="e"
    letter="e"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="f"
    letter="f"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="g"
    letter="g"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="h"
    letter="h"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="i"
    letter="i"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="j"
    letter="j"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="k"
    letter="k"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="l"
    letter="l"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="m"
    letter="m"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="n"
    letter="n"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="o"
    letter="o"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="p"
    letter="p"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="q"
    letter="q"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="r"
    letter="r"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="s"
    letter="s"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="t"
    letter="t"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="u"
    letter="u"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="v"
    letter="v"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="w"
    letter="w"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="x"
    letter="x"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="y"
    letter="y"
  />
  <Connect(LetterGridItem)
    handleGuess={[Function]}
    handleKeyPress={[Function]}
    key="z"
    letter="z"
  />
  <t
    handleEventType="keydown"
    handleFocusableElements={false}
    handleKeys={
      Array [
        "a",
        "b",
        "c",
        "d",
        "e",
        "f",
        "g",
        "h",
        "i",
        "j",
        "k",
        "l",
        "m",
        "n",
        "o",
        "p",
        "q",
        "r",
        "s",
        "t",
        "u",
        "v",
        "w",
        "x",
        "y",
        "z",
      ]
    }
    onKeyEvent={[Function]}
  />
</div>
`;
